File: chapter2/exercises/law_large_numbers.py
# RCDS Introduction to probability and statistical inference.
# Jesús Urtasun Elizari. ICL 2024 / 2025.
# Chapter 2. Expected values and parameter estimation.

# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set up seaborn for prettier plots
sns.set(style = "whitegrid")

# The law of large numbers (LLN)
print("The law of large numbers (LLN)")

# Simulate 5,000 dice rolls. Calculate and store the mean for all samples.
# The running mean, as function of number of samples, should converges to the true mean.

# True means for comparison
dice_mean = 3.5 # True mean for a six-sided die

# Parameters
# num_samples = 50 # Number of samples for LLN
# num_samples = 500 # Number of samples for LLN
num_samples = 5000 # Number of samples for LLN

# Dice Running Mean
dice_rolls = np.random.randint(1, 7, num_samples)  # Simulate 5000 dice rolls
dice_running_mean = np.cumsum(dice_rolls) / np.arange(1, num_samples + 1)

# Plotting LLN results
plt.figure(figsize = (10, 5))

# Plot running mean
plt.plot(dice_running_mean, color = "skyblue", label = "Running Mean")
plt.axhline(dice_mean, color = "red", linestyle = "dashed", linewidth = 2, label = f"True Mean = {dice_mean}")
plt.title("Law of large numbers - Dice roll running mean")
plt.xlabel("Number of rolls")
plt.ylabel("Mean of dice rolls")
plt.legend()
plt.show()

File: chapter2/exercises/law_large_numbers.R
# Load necessary library
library(ggplot2)

# Set seed for reproducibility
set.seed(42)

# The Law of Large Numbers (LLN) .....................................................................
cat("The law of large numbers (LLN)\n")

# Parameters
# num_samples <- 50 # Number of samples for LLN
# num_samples <- 500 # Number of samples for LLN
num_samples <- 5000 # Number of samples for LLN

# True mean for comparison
dice_mean <- 3.5 # True mean for a six-sided die

# Dice Running Mean
dice_rolls <- sample(1:6, num_samples, replace = TRUE) # Simulate 5000 dice rolls
dice_running_mean <- cumsum(dice_rolls) / seq_along(dice_rolls)

# Prepare data for plotting
df <- data.frame(
  RollNumber = 1:num_samples,
  RunningMean = dice_running_mean
)

# Plot running mean
ggplot(df, aes(x = RollNumber, y = RunningMean)) +
  geom_line(color = "skyblue", size = 1) +
  geom_hline(yintercept = dice_mean, color = "red", linetype = "dashed", size = 1) +
  labs(
    title = "Law of Large Numbers - Dice Roll Running Mean",
    x = "Number of Rolls",
    y = "Mean of Dice Rolls"
  ) +
  theme_minimal() +
  annotate("text", x = num_samples * 0.8, y = dice_mean + 0.2, 
           label = paste("True Mean =", dice_mean), color = "red")
